src/protocols/light_client/components/send_last_state_proof.rs

```diff
@@ -179,20 +181,62 @@ impl<'a> SendLastStateProofProcess<'a> {
                     .build();
                 self.nc.reply(self.peer, &message);
             } else {
-                log::warn!("peer {}, build prove request failed", self.peer);
+                warn!("peer {}, build prove request failed", self.peer);
             }
         } else {
+            let reorg_last_headers = headers[..reorg_count]
+                .iter()
+                .map(ToOwned::to_owned)
+                .collect::<Vec<_>>();
+            let mut new_last_headers = headers[reorg_count + sampled_count..]
+                .iter()
+                .map(ToOwned::to_owned)
+                .collect::<Vec<_>>();
+            // Checked in previous steps.
+            assert_eq!(new_last_headers.len(), last_n_count);
+            let last_headers = match last_n_count.cmp(&last_n_blocks) {
+                Ordering::Equal => new_last_headers,
+                Ordering::Greater => {
+                    let split_at = last_n_count - last_n_blocks;
+                    new_last_headers.split_off(split_at)
+                }
+                Ordering::Less => {
+                    if let Some(prove_state) = peer_state.get_prove_state() {
+                        let mut old_last_headers = if reorg_count == 0 {
+                            prove_state.get_last_headers()
+                        } else {
+                            &headers[..reorg_count]
+                        };
+                        // last_headers from previous prove state are empty
+                        // iff the chain only has 1 block after MMR enabled.
+                        if old_last_headers.is_empty() {
+                            new_last_headers
+                        } else {
+                            let required_count = last_n_blocks - last_n_count;
+                            let old_last_headers_len = old_last_headers.len();
+                            if old_last_headers_len > required_count {
+                                let split_at = old_last_headers_len - required_count;
+                                old_last_headers = old_last_headers.split_at(split_at).1;
+                            }
+                            old_last_headers
+                                .iter()
```

Comment:
suggest to use `iterator#skip`

```suggestion
                            old_last_headers
                                .iter()
                                .skip(old_last_headers_len.saturating_sub(required_count))
```